src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,52 @@
+package duke;
+
+import duke.command.*;
+import duke.exception.*;
+import duke.parser.*;
+import duke.ui.*;
+import duke.task.*;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Scanner;
+import java.util.Optional;
+import java.util.NoSuchElementException;
+
+
+public class Duke {
+    public static void main(String[] args) {
+        System.out.printf(Messages.WELCOME);
+
+        Scanner scanner = new Scanner(System.in);
+        List<Task> tasks = new ArrayList<>();
+        Command<List<Task>> currentCommand = null;
+
+        while (true) {
+            try {
+                System.out.printf("\n-> ");
+                String userInput = scanner.nextLine();
+
+                // Parse user input
+                Command<List<Task>> parsedCommand = Parser.parseInput(userInput);
+
+                // Execute the command
+                if (parsedCommand != null) {
+                    currentCommand = parsedCommand;
+                    currentCommand.execute(tasks);
+                }
```

Comment:
The null check seems to be not necessary. This is because instead of nesting your DukeException in Parser, you can bring it up to this level and share this level of try catch, which would simplify the code in the Parser and here. The currentCommand object seems to be a redundant duplicate of parsedCommand too :D